### PR TITLE
Add LE audio connected group

### PR DESCRIPTION
**Pull request by Thalley**
Add the concept of the opaque unicast group. 

For now this is only used for the client to group channels (basically a wrapper around the Connected Isochronous Group (CIG)), but may have applications for the server as well. 

The main usage of the group is when setting the QoS, as all channels in a CIG shall share the same Interval, framing and latency, and the QoS of all channels in a CIG shall be set in a single GATT write command. The update of the QoS function is done in a following PR.

### CODE REVIEW

File: subsys/bluetooth/host/audio/Kconfig.bap [hunk 1]
```diff
@@ -13,15 +13,25 @@ config BT_BAP
...
 	  the unicast client.
 
+config BT_BAP_UNICAST_GROUP_STREAM_COUNT
+	int "Basic Audio Unicast Group Stream count"
+	depends on BT_BAP_UNICAST_GROUP_COUNT > 0
+	default 1
+	range 0 BT_ISO_MAX_CHAN if BT_ISO_MAX_CHAN < 31
+	range 0 31
+	help
+	  This option sets the maximum number of streams per broadcast source
```
Comment (asbjornsabo) [inline subsys/bluetooth/host/audio/Kconfig.bap:32]: This help text does not seem correct - it talks about "broadcast".
I assume this would be the number of streams per (unicast) group.

File: subsys/bluetooth/host/audio/chan.c [hunk 2]
```diff
@@ -1023,14 +1023,20 @@ int bt_audio_cig_terminate(struct bt_audio_chan *chan)
...
 		struct bt_iso_cig *cig;
 
-		cig = cigs[i];
+		cig = unicast_groups[i].cig;
 		if (cig != NULL &&
 		    cig->cis != NULL &&
 		    cig->cis[0] == chan->iso) {
-			return bt_iso_cig_terminate(cig);
+			err = bt_iso_cig_terminate(cig);
+			if (err == 0) {
```
Comment (asbjornsabo) [inline subsys/bluetooth/host/audio/chan.c:1035]: (Seems like an unrelated bugfix.  Next time, I would prefer that as a separate commit.  OK for now.)

File: include/bluetooth/audio.h [hunk 3]
```diff
@@ -968,6 +971,23 @@ int bt_audio_chan_unlink(struct bt_audio_chan *chan1,
  */
 int bt_audio_chan_send(struct bt_audio_chan *chan, struct net_buf *buf);
 
+/** @brief Create audio unicast group.
+ *
+ *  Create a new audio unicast group with one or more audio channels as a
+ *  cental. Channels in a unicast group shall share the same interval, framing
```
Comment (asbjornsabo) [inline include/bluetooth/audio.h:977]: "central" (missing "r")

File: include/bluetooth/audio.h [hunk 4]
```diff
@@ -968,6 +971,23 @@ int bt_audio_chan_unlink(struct bt_audio_chan *chan1,
...
+ *
+ *  @param[in]  chans          Array of channel objects being used for the
+ *                             group.
+ *  @param[in]  num_chan       Number of channels in @p chans.
+ *  @param[out] unicast_group  Pointer to the unicast group created
+ *
+ *  @return Zero on success or (negative) error code otherwise.
+ */
+int bt_audio_unicast_group_create(struct bt_audio_chan *chans,
+				  uint8_t num_chan,
```
Comment (asbjornsabo) [inline include/bluetooth/audio.h:988]: Alignment seems off - but that may just be Github

File: include/bluetooth/audio.h [hunk 5]
```diff
@@ -988,6 +988,17 @@ int bt_audio_unicast_group_create(struct bt_audio_chan *chans,
 				  uint8_t num_chan,
 				  struct bt_audio_unicast_group **unicast_group);
 
+/** @brief Delete audio unicast group.
+ *
+ *  Delete a audio unicast group as a central. All channels in the group shall
+ *  be in an idle or configured state.
```
Comment (asbjornsabo) [inline include/bluetooth/audio.h:994]: Proposal: "in idle state or configured state".  (I think the "an" should be removed.)

File: subsys/bluetooth/host/audio/chan.c [hunk 6]
```diff
@@ -1219,6 +1219,31 @@ int bt_audio_unicast_group_create(struct bt_audio_chan *chans,
...
+
+	CHECKIF(unicast_group == NULL) {
+		BT_DBG("unicast_group is NULL");
+		return -EINVAL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&unicast_group->chans, chan, node) {
+		if (chan->state != BT_AUDIO_CHAN_IDLE &&
+		    chan->state != BT_AUDIO_CHAN_CONFIGURED) {
+			BT_DBG("chan %p is not idle", chan);
```
Comment (asbjornsabo) [inline subsys/bluetooth/host/audio/chan.c:1234]: Proposal: "[...] state is neither idle nor configured"

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
Comments from review of the first three commits.  (Have not reviewed the test yet.)

File: subsys/bluetooth/host/audio/chan.c (hunk 2 above)
Comment (Thalley) [reply subsys/bluetooth/host/audio/chan.c:1035]: It is somewhat unrelated, so could have been a different commit.

File: include/bluetooth/audio.h (hunk 4 above)
Comment (Thalley) [reply include/bluetooth/audio.h:988]: Looks correct in my editor at least :D

File: include/bluetooth/audio.h (hunk 5 above)
Comment (Thalley) [reply include/bluetooth/audio.h:994]: Replaced "an" with "the"

File: subsys/bluetooth/host/audio/chan.c (hunk 6 above)
Comment (Thalley) [reply subsys/bluetooth/host/audio/chan.c]: Made it a bit more generic and added logging of the state that is invalid.

**Review by asbjornsabo — CHANGES_REQUESTED** [review]
Some questions and comments, and a couple of changes that may be necessary

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+					struct bt_audio_ep *ep,
+					struct bt_audio_capability *cap,
+					struct bt_codec *codec)
+{
+	printk("ASE Codec Config: conn %p ep %p cap %p\n", conn, ep, cap);
+
+	print_codec(codec);
+
+	for (size_t i = 0; i < ARRAY_SIZE(g_channels); i++) {
+		if (g_channels[i].conn == NULL) {
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:90]: The easiest (only?) way of checking if a channel is free, is to check the `conn` pointer at the moment. 

The connection pointer will be set by the stack.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+CREATE_FLAG(flag_chan_qos);
+CREATE_FLAG(flag_chan_enabled);
+
+static void print_codec(const struct bt_codec *codec)
+{
+	printk("codec 0x%02x cid 0x%04x vid 0x%04x count %u\n",
+	       codec->id, codec->cid, codec->vid, codec->data_count);
+
+	for (uint8_t i = 0; i < codec->data_count; i++) {
+		printk("data #%zu: type 0x%02x len %u\n",
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:52]: Typo :) Initialized I had used `size_t` for the loop variable, but decided to use the type of the value we are comparing it against instead.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+
+static int lc3_reconfig(struct bt_audio_chan *chan,
+			struct bt_audio_capability *cap,
+			struct bt_codec *codec)
+{
+	printk("ASE Codec Reconfig: chan %p cap %p\n", chan, cap);
+
+	print_codec(codec);
+
+	SET_FLAG(flag_chan_configured);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:108]: `lc3` here refers to the capability that was registered that uses these callback functions (i.e. a capability that uses the LC3 codec, hence the naming). 

The naming was taken from the BAP shell module, but it may be better to rename the callback functions to something else.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+		capability_found = true;
+		return;
+	}
+
+	if (ep != NULL) {
+		if (params->type == BT_AUDIO_SINK) {
+			add_remote_sink(ep, params->num_eps);
+			endpoint_found = true;
+		} else {
+			FAIL("Invalid param type: %u\n", params->type);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:253]: We only do discovery for sinks in this test at this point :)

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+		}
+
+		return;
+	}
+
+	printk("Discover complete: err %d\n", params->err);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:259]: > errwill alway be zero here, correct? (Due to the return in line 239.)

Yes, will remove the logging of that

> Also, this line will only be executed if the endpoint was NULL, due to the return in 256. Should that return be grouped with the "else"?

No, discovery will be completed when both `cap` and `ep` are NULL. 

> Finally: Would it be better to have this line at the very end of the function, after all is done, or with the SET_FLAG? As it currently is, the function can print "Discover complete", and still FAIL.

Discovery can still complete successfully without finding any endpoints or capabilities, so I would argue no.

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err != 0) {
+		FAIL("Failed to connect to %s (%u)\n", addr, err);
+
+		bt_conn_unref(conn);
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_client_test.c:291]: The lines actually hardly matters since the test fails, but I've moved them regardless :)

File: tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_server_test.c [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+
+#if defined(CONFIG_BT_BAP) && defined(CONFIG_BT_AUDIO_BROADCAST)
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/audio.h>
+#include "common.h"
+
+extern enum bst_result_t bst_result;
+
+#define LC3_PRESET(_name, _codec, _qos) \
```
[the remark this answers is not in the document]
Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_audio/src/unicast_server_test.c:15]: Good idea, added

**Review by asbjornsabo — APPROVED** [review]
Seems good to go.
